staging/src/k8s.io/apiserver/plugin/pkg/audit/webhook/webhook.go

```diff
@@ -0,0 +1,215 @@
+	// round trip to the external audit service before sending a response.
+	ModeBlocking = "blocking"
+)
+
+// AllowedModes is the modes known by this webhook.
+var AllowedModes = []string{
+	ModeBatch,
+	ModeBlocking,
+}
+
+const (
+	// Default configuration values for ModeBatch.
+	//
+	// TODO(ericchiang): Make these value configurable. Maybe through a
+	// kubeconfig extension?
+	defaultBatchBufferSize = 1000        // Buffer up to 1000 events before blocking.
+	defaultBatchMaxSize    = 100         // Only send 100 events at a time.
+	defaultBatchMaxWait    = time.Minute // Send events at least once a minute.
+)
+
+var (
+	// NOTE: Copied from other webhook implementations
+	groupFactoryRegistry = make(announced.APIGroupFactoryRegistry)
+	groupVersions        = []schema.GroupVersion{auditv1alpha1.SchemeGroupVersion}
+	registry             = registered.NewOrDie("")
+)
+
+func init() {
+	registry.RegisterVersions(groupVersions)
+	if err := registry.EnableVersions(groupVersions...); err != nil {
+		panic(fmt.Sprintf("failed to enable version %v", groupVersions))
+	}
+
+	install.Install(groupFactoryRegistry, registry, audit.Scheme)
+}
+
+// AuditBackend is a webhook implementation of the audit backend.
+type AuditBackend struct {
+	w *webhook.GenericWebhook
+
+	// buffer, if non-nil, indicates that the backend should operate in
+	// batch mode, buffering a set of events and batching them in one
+	// requests to the backend.
+	buffer chan *auditinternal.Event
+	// Maximum number of events that can be sent at once.
+	maxBatchSize int
+	// Amount of time to wait after sending events before force sending another set.
+	//
+	// Receiving maxBatchSize events will always trigger a send, regardless of
+	// if this amount of time has been reached.
+	maxBatchWait time.Duration
+}
+
+// NewBackend returns an audit backend at sends events over HTTP to an external service.
+// The mode indicates the caching behavior of the webhook. Either blocking (ModeBlocking)
+// or buffered with batch POSTs (ModeBatch).
+func NewBackend(kubeConfigFile string, mode string) (*AuditBackend, error) {
+	w, err := webhook.NewGenericWebhook(registry, audit.Codecs, kubeConfigFile, groupVersions, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	b := &AuditBackend{w: w}
+
+	switch mode {
+	case ModeBatch:
+		// enable batching.
+		b.buffer = make(chan *auditinternal.Event, defaultBatchBufferSize)
+		b.maxBatchSize = defaultBatchMaxSize
+		b.maxBatchWait = defaultBatchMaxWait
+	case ModeBlocking:
+	default:
+		return nil, fmt.Errorf("webhook mode %q is not in list of known modes (%s)",
+			mode, strings.Join(AllowedModes, ","))
+	}
+
+	return b, nil
+}
+
+// Run starts the webhook and shuts down once stopCh closes. It MUST be called before
+// using the backend. Run is non-blocking.
+func (a *AuditBackend) Run(stopCh <-chan struct{}) error {
+	if a.buffer == nil {
+		// Not buffered, nothing to do.
+		return nil
+	}
+
+	go func() {
+		for {
```

Comment:
I think you should add a `defer runtime.HandleCrash()` somewhere. One option is to wrap this in a `wait.Until(func, 0 /* period */, stopCh)`, but I'm not sure what the implications of a 0-duration timer are.